### PR TITLE
feat: improve tools to include name and add tests

### PR DESCRIPTION
**Pull request by drbh**
This PR makes tool calling aware of the name of the function selected. 

Fixes: https://github.com/huggingface/text-generation-inference/issues/1657

Thank you @puppetm4st3r for the helpful snippets, large parts of this PR are simply refactors of the code shared 🙏

**opening draft PR because small tweaks are needed before merging

### CODE REVIEW

**Comment by puppetm4st3r** [comment]
thanks to you, you are doing a great job with TGI, it is an honor to be able to contribute with a little grain of sand.

I have looked the changes on this PR, and I can comment that during my local tests, the tool prompt greatly influences the final performance of the tool selection or error notification when none can be selected, adding -------------{}{} ----------- delimiting to build the final prompt influences very heavy on the results, so I recommend considering format the final prompt delimited like this:

```
// Proceed only if tool_prompt is not None
     if let Some(tool_prompt) = &req.tool_prompt {
         // Find the last message with role 'user'
         if let Some(last_user_message) = req.messages.iter_mut().rev().find(|msg| msg.role == "user") {
             // Generate the additional content combining tool_prompt and tools_str
             let additional_content = format!("\n\n---------------------------\n{}{}\n---- -----------------------", tool_prompt, tools_str);
            
             // If the last user message has existing content, append the additional content to it
             if let Some(content) = &mut last_user_message.content {
                 content.push_str(&additional_content); // Append to the existing content
             } else {
                 // If, for some reason, the content is None, replace it with the additional_content
                 last_user_message.content = Some(additional_content);
             }
         }
         // Note: If there is no message with role 'user', this block does nothing
     }
```

also changing the prompt from lib.rs
```
fn default_tool_prompt() -> Option<String> {
    Some(
        "\nBased on the conversation, please choose the most appropriate tool to use: ".to_string(),
    )
}
```

to:

`You will be presented with a JSON schema representing a set of tools.\nIf the user request lacks of sufficient information to make a precise tool selection: Do not invent any tool's properties, instead notify with an error message.\ n\nJSON Schema:\n
`

has a positive impact in the performance of the llm on the tool selection task.

to get the final prompt delimited I have to change the order of processing in the request, in the original code the tokenizer applies very on the first fragments on the code, to build the delimited final prompt I have to move the tokenizer process after the grammar generation and function definitions in order to build a final prompt with all stuff ready before tokenization process.

**Comment by drbh** [comment]
Hi @puppetm4st3r thank you for the feedback on the PR , I've updated the default prompt and looked into changing how the tools are appended to the input.

Currently I'm not 100% sure what approach is best for passing tools to the LLM. 

At the moment we send:

`{inputs}{tool_prompt}{tools_str}`

where input is the users input after the template is applied to the incoming messages.

The method proposed in your PR seems to append the `{tool_prompt}{tools_str}` to the last message, and then applies the chat template.

Its unclear to me which approach is best and would love to find an alternative that enables this flexibility without adding more complexity (an extra template)...

**Comment by puppetm4st3r** [comment]
yes, that part is a bit complex in Use cases, my motivation was to append to the last user message in order to try not confuse the llm, also implemented de callId property of the open ai spec because is useful in chathistory for the llm to understand previous turns of conversations that have called a function, the llm have to undestand and distinguish between turns funcion calls to not get confused,  Honestly, I didn't give much thought to what was the best strategy to include the prompt_tool and the tool definitions, I simply tried to maintain the consistency of messaging with the llm so as not to confuse it and I chose to modify the user's last message since that modification It is ephemeral and does not come back to the user (it is like a patch, I don't like it but I didn't find another direct and simple way to implement it by the moment)

**Comment by puppetm4st3r** [comment]
Another option that I test a few months ago is to build a wrapper for the official OPEN AI client, which preset the chat template with the required stuff to execute function calling using json grammar guidance, the benefit you don't touch the server, the cons is that you create a new dependency/artifact to call functions and that breaks the concept of open ai compatible api., is more like open ai compatible client.

this option I feel takes us away from the original purpose

**Comment by drbh** [comment]
@puppetm4st3r I think for now we'll include the current changes and keep the discussion open regarding tools and the template. It seems like there may be some convergence in using tools within templates https://github.com/huggingface/transformers/blob/76fa17c1663a0efeca7208c20579833365584889/src/transformers/models/cohere/tokenization_cohere_fast.py#L292 and this would require a larger change to the template logic.

**Comment by puppetm4st3r** [comment]
great, yes Im ok that refactors on my locals with something like that at chat template level, now Im on the gym, later ar night I can share my chat templates that have result in good llm tool understansing on chat history and large multi turn conversarions, will back with more detailed ideas later! c ya

**Comment by puppetm4st3r** [comment]
continuing the refinement of the idea,

Let's review the key findings we have uncovered during the implementation of this feature

1. Models are highly sensitive to the chat template, with some templates not performing well with sentences containing non-alphanumeric characters, such as a JSON schema. Through experimentation, the "ChatML" template was found to be the most effective.

3. Models not natively trained for function calling, attempting to mimic this behavior, are constrained to generating a JSON schema structure, hence incapable of returning feedback or non tool selection natural text. To mitigate this, the previous PR added the capability for the LLM to communicate errors (appending a "system" tool in a non clear or straightforward way), which is some crucial as this kind of models will not support optional use of tools.

5. Due to the sensitivity to chat templates, the user’s prompt must include instructions for the model to know which JSON schema to use for text generation. Without this, the generation process could be significantly delayed or result in empty tokens until the context window is filled.

7. Effective prompt engineering significantly enhances the model's tool selection quality, with even error notification potentially failing if the prompt is unclear or does not adhere to the chat template.
8. For chat history, only ChatML considers roles for tools (representing a tool call by the LLM), where the call ID is essential for the LLM to understand and logically sequence the calls and responses from the selected tools. However, in models not trained to understand the tool role, a workaround has been to modify the Jinja template of the chat template to render the tool call as part of the Assistant role, like "I have decided to call the following tool: …".

Conclusión:
Incorporating tool functionality into the TGI API presents significant challenges, particularly in maintaining clean integration without resorting to less desirable practices like modifying the last user role message. The interplay between chat templates, prompt engineering, and model capabilities is crucial for effective function calling, necessitating careful design and customization to achieve seamless and efficient tool integration and communication.

Thats for now, we will need to keep double clicking on these ideas and concepts.

**Comment by puppetm4st3r** [comment]
ps: in my use case that work for me was modify the chatML template in order to support tools for non-pretrained LLM for tool call understanding plus appending the prompt_tools and prompt_str to the last user's role message, thats work fine, but it is not the most direct and clean solution. We now whats work but we need to implement that in a clear and best practice way.

this is my chatML modified template:
```
{% for message in messages %}
  {%- if message['role'] == 'system' %}
<|im_start|>system\n{{ message['content'] }}<|im_end|>\n
  {%- elif message['role'] == 'user' %}
<|im_start|>user\n{{ message['content'] }}<|im_end|>\n
  {%- elif message['role'] == 'tool' %}
<|im_start|>tool\ncall_id: {{ message['tool_call_id'] }}\nresult: {{ message['content'] }}<|im_end|>\n
  {%- elif message['role'] == 'assistant' %}
<|im_start|>assistant{%- if message.tool_calls -%}
      {%- for tool_call in message.tool_calls %}
tool_execution: call_id: {{ tool_call.id }}, function: {{ tool_call.function.name }}, arguments: {{ tool_call.function.arguments }}
      {%- endfor -%}
      {%- if not loop.last or not add_generation_prompt %}<|im_end|>{% endif -%}
    {%- else %}
{{message['content'] }}{%- if not loop.last or not add_generation_prompt %}<|im_end|>{% endif -%}
    {% endif -%}
  {% endif -%}
{% endfor -%}
{% if add_generation_prompt -%}
<|im_start|>assistant
{% endif %}
```

File: router/src/server.rs [hunk 1]
```diff
@@ -754,19 +754,34 @@ async fn chat_completions(
...
+        seed,
+        stop,
+        stream,
+        temperature: _,
+        tools,
+        tool_choice,
+        tool_prompt,
+        top_p: _,
+        top_logprobs: _,
+    } = req;
```
Comment (Narsil) [inline router/src/server.rs:775]:
```suggestion
   ..
   }  = req;
   ```

Comment (Narsil) [reply router/src/server.rs:775]: You can remove all the `_` using `..`.

**Pull request merged**